src/modules-by-example.adoc

```diff
@@ -0,0 +1,320 @@
+}
+----
+
+Here’s an Ion 1.1 document that’s equivalent, in the sense that it allocates symbol IDs in the
+same order. (The IDs will be different, though, due to new system symbols.)
+
+[{nrm}]
+----
+*$ion_1_1*
+**$ion_encoding**::{
+  modules: [
+    (*module* local
+      (*symbols* [ "s1", "s2" ]))
+  ],
+  install_symbols: [ local ]
+}
+----
+
+The definition of the local symbol table has been refactored into two parts. First, the list of
+symbols is expressed inside a module named `local`. Then, the symbols from that module are
+installed to form the new local symbol table. Compared to the behavior of `$ion_symbol_table`,
+this is akin to defining a named symbol table “inline” to hold local symbols, then defining the
+local symbol table only via `imports` and no `symbols` field.
+
+Let's look more closely at the definition of `local`:
+
+[{nrm}]
+----
+(*module* local
+  (*symbols* ["s1", "s2"]))
+----
+
+The `*module*` keyword starts an S-expression that defines a new module with the given name.
+The `*symbols*` keyword starts a subform that defines the module's exported symbol table.
+This clause expects a list of strings, and both the syntax and semantics are the same as the
+`symbols` field of `$ion_shared_symbol_table`.
+
+Once this module is defined, we can _install_ it into the local symbol table:
+
+[{nrm}]
+----
+  install_symbols: [ local ]
+----
+
+This field expects a list of symbols that match names declared in the `modules` field.  The
+resulting local symbol table is simply the concatenation of the exported symbol tables of those
+modules.  This works the same way as the `imports` field of `$ion_symbol_table`.
+
+
+=== Importing Symbols
+
+Given the equivalencies above, we could perform a naive round-trip of the preceding 1.1 document
+back to 1.0. First, turn the `local` module into the equivalent shared symbol table:
+
+[{nrm}]
+----
+**$ion_shared_symbol_table**::{
+  name: "local",
+  version: 1,
+  symbols: ["s1", "s2"]
+}
+----
+
+Then translate `install_symbols:[local]` into its 1.0 equivalent:
+
+[{nrm}]
+----
+*$ion_1_0*
+**$ion_symbol_table**::{
+  imports: [{ name: "local", version: 1, max_id: 2 }]
+}
+----
+
+NOTE:  Even ignoring Ion 1.1, this is how you would hoist local symbols into a
+new shared symbol table.
+
+The latter imports-only document has this 1.1 equivalent:
+
+
+[{nrm}]
+----
+*$ion_1_1*
+**$ion_encoding**::{
+  modules: [
+    (*import* local "local" 1 2)
+  ],
+  install_symbols: [ local ]
+}
+----
+
+Here we see a new form inside the `modules` field that imports a module into the encoding
+environment and assigns it a name.
+The `*import*` keyword starts an S-expression that expects three or four arguments. The first is
+a symbolic name that we can use later to refer to the imported module.  The remaining arguments
+are effectively the `name`, `version` and `max_id` fields of the 1.0 `imports` struct.
+
+TIP: From the perspective of Ion 1.1, shared symbol tables _are_ encoding modules.
```

Comment:
+1

I was going to ask what is the significance of this fact—can an Ion 1.1 encoding directive import an Ion 1.0 shared symbol table?

Reply:
Yes--but more importantly a few things:

* Shared symbol tables in Ion 1.1 ***are*** shared encoding modules.
* The syntax for shared symbol tables is extended to support macros in Ion 1.1 (we don't have a convention for something like `$ion_shared_module::{...}`
* Ion 1.0 local symbol tables ***are*** encoding directives.

I think we need to be super crisp about these facts (or correct my understanding if these facts are wrong).

Reply:
Yes, an Ion 1.1 encoding directive can import an Ion 1.0 shared symbol table. That's exactly what's shown by the preceding example.

Reply:
Hmm, then I do think we need some clarifying text here--because as it is the dissonance between Ion 1.0 local symbol tables, Ion 1.0/1.1 shared symbol tables (shared modules?), and Ion 1.1 encoding directives are muddled to me, and part of that muddling is the recycling of the shared symbol table schema (which practically makes sense, but really hurts my brain around the mental model)

Reply:
I think this would be a great topic for our meetup next week.